API 2.0

avive opened the pull request:
Draft PR to review and discuss GlobalState service changes to support Spacemesh Apps in release 0.3.

YaronWittenstein commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,124 @@
+}
+
+// Information about an app template in the global state store
+message AppTemplate {
```
I think we can make it `Template`

avive replied:
Done - and same everywhere

YaronWittenstein commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,124 @@
+}
+
+// The address of an app template in the templates global state store
+message AppTemplateAddress {
```
In the SVM code there is no `AppTemplateXXX` - so I'd rename it to `TemplateAddress`

avive replied:
done

YaronWittenstein commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,124 @@
+    uint32 id = 1; // var index in app's state
+    string name = 2; // var name
+    VariableType type = 3; // type
+    bool is_array = 4; // when true, variable is an array of 0 or more primitive value types, otherwise, a single primitive type value.
```
I'm not well versed in Profobuf - but if it's possible to specify that a variable is either a primitive or an Array of primitives I would do that.

YaronWittenstein replied:
In the future, there might be more ways to represent non-primitive variables...

avive replied:
done

YaronWittenstein commented on proto/spacemesh/v1/global_state.proto:
```diff
@@ -2,14 +2,28 @@ syntax = "proto3";
+    rpc GetApp(GetAppRequest) returns (GetAppResponse);
+
+    // Get an app's variable metadata and optionally value
+    rpc GetAppVariable(GetAppVariableRequest) returns (GetAppVariableResponse);
```
As I've said - I wouldn't add it right now.
Probably it won't be used when we have only fixed-sized variables.

avive replied:
let's talk about - i think it will be very useful for api clients even with fixed-size vars

YaronWittenstein commented on proto/spacemesh/v1/global_state.proto:
```diff
@@ -2,14 +2,28 @@ syntax = "proto3";
     rpc GlobalStateHash(GlobalStateHashRequest) returns (GlobalStateHashResponse);
 
+    // Get an app template meta-data and code
+    rpc GetAppTemplate(GetAppTemplateRequest) returns (GetAppTemplateResponse);
```
`GetTemplate` / `GetTemplateResponse` ?

YaronWittenstein commented on proto/spacemesh/v1/global_state.proto:
```diff
@@ -2,14 +2,28 @@ syntax = "proto3";
+    rpc GetAppTemplate(GetAppTemplateRequest) returns (GetAppTemplateResponse);
+
+    // Get all app templates
+    rpc GetAllAppTemplates(GetAllAppTemplatesRequest) returns (GetAllAppTemplatesResponse);
```
Since we want to support pagination here, I prefer to call it `GetTemplates`, or maybe `TemplateIterator`

for the response type - how about `GetTemplatesResponse` ?

YaronWittenstein commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,124 @@
+}
+
+// A response with AppTemplate data
+message GetAppTemplateResponse {
```
`GetTemplateResponse` ?

YaronWittenstein commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,124 @@
+}
+
+// A request to return a template by address from the template global state store
+message GetAppTemplateRequest {
```
`GetTemplateResquest` ?

YaronWittenstein commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,124 @@
+message GetAppRequest {
+    AccountId address = 1; // app address
+    LayerNumber layer_number = 2; // layer for state. Ignored when include_var_values is set to false. pass MAX_UI32 to get latest state.
+    bool include_var_values = 3; // when true, var values will be included in response, false otherwise.
```
I didn't add that option under the issue - but I do think it makes sense to have it

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+
+// The meta-data and value of an app's variable
+message Variable {
+    VariableMetadata meta_data = 1;
```
```suggestion
    VariableMetadata metadata = 1;
```

avive replied:
done

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+    repeated bytes values = 1;
+}
+
+// The meta-data and value of an app's variable
```
```suggestion
// The metadata and value of a state value
```

avive replied:
done

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+}
+
+// The address of an app template in the templates global state store
+message TemplateAddress {
```
Why do we need a distinct type for this? Why not use the existing `Address` type?

avive replied:
I think because in current SVM design, templates are stored in their own namespace which is not just an abstracted account address. We reviewed this as far as I remember in recent svm calls. Yaron can clarify.

avive replied:
Item for design review zoom

lrettig replied:
then, rather than introducing a new type, maybe add a `namespace` to the existing `Address` type

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+// Information about an app template in the global state store
+message Template {
+    TemplateAddress address = 1; // template address
+    string name = 2; // template name
```
See comment in issue, I don't understand why this is required

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+    string name = 2; // template name
+
+    // note - the author can be obtained from the deploy transaction but it is nice to have it here
+    AccountId author = 3; // template uploader
```
```suggestion
    AccountId deployer = 3; // template uploader
```

avive replied:
I'm using the svm terminology as specified by yaron, until we discuss making any changes to this terminology with yaron.

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -50,22 +50,42 @@ message AccountDataQueryRequest {
     uint32 offset = 3; // return results from offset
 }
 
+message ExecutionError {
```
We have standard error/status types already, with the same data elements, maybe we don't need to introduce another one here. See "google.golang.org/grpc/codes", "google.golang.org/grpc/status"

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+// but is convenient to return them directly as part of apps' metadata.
+message App {
+    AccountId address = 1; // App's address in global space
+    string name = 2; // short descriptive name
```
Do we need this?

avive replied:
I think it is good to have a short app name that describes it. I think that it is something that is set by spwaner and spawn time per app.

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+    AccountId address = 1; // App's address in global space
+    string name = 2; // short descriptive name
+    TemplateAddress template_address = 3; // App's template address
+    LayerNumber spawn_layer = 4; // Layer in which app was spawned
```
Why is this needed?

avive replied:
Which field? These are all meta-data about an app which is useful. It is useful to know in which layer an app was deployed to the mesh although this info can be obtained from the spawn transaction - it is just convince to include some metadata that can be obtained from transactions about the app in response to this call because the intent of the caller is to get this data.

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+    repeated Template templates = 2;
+}
+
+// A request to get an app's variable by var id
```
Instead of "variable" can we give these a more descriptive name, e.g., state entry?

lrettig replied:
@YaronWittenstein said these are called "storage variables" in SVM

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+// A request to get an app's variable by var id
+message GetVariableRequest {
+    AccountId address = 1; // app's address
+    LayerNumber layer_number = 2; // requested value as of a layer. Pass MAX_UI32 to get latest value
```
As noted in the issue, will only work for archival nodes! What do non-archival nodes do if this is specified? Error?

avive replied:
See my comment there - full nodes will hold state for n layers back and archive nodes since genesis but from an api level it is good to have this like this and nodes will return not_found status if they can't satisfy this specific request for a layer which is not recent.

lrettig replied:
I think that works, but for the record, we cannot force full nodes to hold intermediate state for any minimum number of layers since this is extraprotocol

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,129 @@
+message GetVariableRequest {
+    AccountId address = 1; // app's address
+    LayerNumber layer_number = 2; // requested value as of a layer. Pass MAX_UI32 to get latest value
+    uint32 id = 3; // app's var id - var's unique identifier
```
Where does the caller look this up?

avive replied:
Caller has the app's schema which includes the variable ids.

avive commented:
@lrettig @YaronWittenstein - I've updated the api to reflect the design review session feedback. I still believe we should have a strongly typed GetVariable() method which is very useful for quick access to app state by dapps. Perhaps we can proceed before another review session by you guys just reviewing the most recent changes? The relent types are in app_types.proto, var_types.proto and global_state.proto

lrettig commented on proto/spacemesh/v1/api_config.yaml:
```diff
@@ -103,7 +115,7 @@ http:
       body: "*"
     - selector: spacemesh.v1.SmesherService.StopPostDataCreationSession
-      post: "/v1/smesher/stoppostdatacreationsession"
+      post: "/v1/smesher/stoppost:datacreationsession"
```
is the `:` a typo?

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,111 @@
+    TemplateAddress address = 1; // template address
+    // note - the author can be obtained from the deploy transaction but it is nice to have it here
+    AccountId author = 2; // template uploader
+    LayerNumber deploy_layer = 3; // layer that the template was deployed - 0 for genesis templates
```
is this actually stored in the mesh data for a template? is it necessary for SVM? (both layer, and txid)

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,111 @@
+
+// Template code
+message Code {
+    string target_svm_version = 4; // at deploy time
```
Can we collapse target svm version and kind down into one field? Can't it be included in the code itself as a "magic word" (first few bytes)?

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,111 @@
+
+// Information about a template in the global state store
+message Template {
+    DeployData deploy_data = 1; // always provided.
```
is this included in mesh data? is it required?

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,111 @@
+message GetTemplatesRequest {
+    uint32 max_results = 1; // max number of results to return
+    uint32 offset = 2; // query offset
+    bool include_code = 3; // include code in response
```
why not just always include both code and metadata? is it any additional work on the backend? it's not that much additional data

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,111 @@
+    bool include_meta_data = 4; // include metadata in response if provided by author
+}
+
+// A response includes total templates available in store and a list of zero ore more templates
```
```suggestion
// A response includes total templates available in store and a list of zero or more templates
```

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -0,0 +1,111 @@
+    AccountId address = 1; // App's address in global space
+    string name = 2; // short descriptive name (optional)
+    TemplateAddress template_address = 3; // App's template address
+    LayerNumber spawn_layer = 4; // Layer in which app was spawned
```
as with templates, above: are spawn layer, spawn txid, and spawner really needed for SVM/stored in mesh?

lrettig replied:
as discussed today, these also belong in metadata, as they're not required by SVM. they can optionally be automatically populated at deploy time if the deployer requests this, and pays for it.

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
+    uint32 log_level = 1; // info, debug, trace, error, etc...
+    string context = 2; // function name e.g. withdraw()
+    string name = 3; // entry name
+    string entry = 4; // entry printed data
```
in EVM/solidity, pretty sure this is a repeated set of field names and values rather than a single blob (per context/name)

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
         TRANSACTION_RESULT_INSUFFICIENT_GAS = 4; // out of gas
         TRANSACTION_RESULT_INSUFFICIENT_FUNDS = 5; // failed due to sender's insufficient funds
+        TRANSACTION_RESULT_APP_NOT_FOUND = 6; // an attempt to execute an app where there's no app at the app address
+        TRANSACTION_RESULT_FUNC_NOT_FOUND = 8; // called function not found in app
```
is this distinguished, at the SVM level, from any other simple execution error? how/why?

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
         TRANSACTION_RESULT_INSUFFICIENT_FUNDS = 5; // failed due to sender's insufficient funds
+        TRANSACTION_RESULT_APP_NOT_FOUND = 6; // an attempt to execute an app where there's no app at the app address
+        TRANSACTION_RESULT_FUNC_NOT_FOUND = 8; // called function not found in app
+        TRANSACTION_RESULT_INSTANTIATION_FAILED = 9; // error creating an instance from a template
```
call this spawn rather than instantiation for consistency

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
+        TRANSACTION_RESULT_APP_NOT_FOUND = 6; // an attempt to execute an app where there's no app at the app address
+        TRANSACTION_RESULT_FUNC_NOT_FOUND = 8; // called function not found in app
+        TRANSACTION_RESULT_INSTANTIATION_FAILED = 9; // error creating an instance from a template
+        TRANSACTION_RESULT_DEPLOY_FAILED = 10; // error deploying a template to the mesh
```
not sure we need to break out these two failure types for spawn/deploy since it should be clear which it is from context (tx type), no?

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -50,22 +50,46 @@ message AccountDataQueryRequest {
+    uint32 index = 6; // the index of the tx in the ordered list of txs to be executed by stf in the layer
+
+    bytes svm_data = 7; // SVM data returned from an SVM function. Needs to be decoded with svm-codec
+    bytes app_state = 8; // when app function executed or app deployed, the app root state hash after execution has completed
```
please call this `state_root` or something to be clearer

lrettig commented on proto/spacemesh/v1/types.proto:
```diff
@@ -74,6 +74,7 @@ message Activation {
     AccountId coinbase = 4; // coinbase account id
     ActivationId prev_atx = 5; // previous ATX pointed to
     uint64 commitment_size = 6; // commitment size in bytes
+    EpochNumber epoch = 7; // Activation on mesh epoch number (when layer is not used)
```
I don't understand the comment here, when is "layer not used"?

lrettig commented on proto/spacemesh/v1/types.proto:
```diff
@@ -110,9 +111,11 @@ message Block {
+        LAYER_STATUS_UNSPECIFIED = 0;
         LAYER_STATUS_APPROVED = 1; // approved by hare
         LAYER_STATUS_CONFIRMED = 2; // confirmed by tortoise
+        LAYER_STATUS_PENDING = 3; // not yet approved, confirmed or verified
```
shouldn't pending come before approved and confirmed?

lrettig commented on proto/spacemesh/v1/var_types.proto:
```diff
@@ -0,0 +1,59 @@
+syntax = "proto3";
```
per our conversation last week, I don't think any of this belongs in the L1 API. I think it belongs at the application layer, and that this data should be returned from this API as a blob of "SVM compatible data"

avive commented:
@YaronWittenstein @lrettig - please check the code - I think I've incorporated everything we talked about and removed non-app related changes from this pr.

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -2,47 +2,32 @@ syntax = "proto3";
-    Code code = 2; // template wasm code. Provided if caller specified to get code.
-    bytes metadata = 3; // storage layout, api, header. Provided if caller specified to get metadata and template include metadata.
+    Code code = 1; // template wasm code. Provided if caller specified to get code.
+    bytes metadata = 2; // storage layout, api, header. Provided if caller specified to get metadata and template include metadata.
```
```suggestion
    bytes metadata = 2; // storage layout, api, header. Provided if caller specified to get metadata and template including metadata. Caller is responsible for decoding.
```

lrettig commented on proto/spacemesh/v1/app_types.proto:
```diff
@@ -2,47 +2,32 @@ syntax = "proto3";
     TEMPLATE_DATA_FLAGS_UNSPECIFIED = 0;
     TEMPLATE_DATA_FLAGS_CODE = 1;
     TEMPLATE_DATA_FLAGS_METADATA = 2;
+    // may need deploy data section - tbd - based on deploy tx format. Understand eth one.
```
```suggestion
    // TODO: may need deploy data section - tbd - based on deploy tx format. Understand eth one.
```

lrettig commented on proto/spacemesh/v1/global_state_types.proto:
```diff
@@ -70,14 +70,10 @@ message TransactionReceipt {
-        TRANSACTION_RESULT_INSTANTIATION_FAILED = 9; // error creating an instance from a template
-        TRANSACTION_RESULT_DEPLOY_FAILED = 10; // error deploying a template to the mesh
+        TRANSACTION_RESULT_RUNTIME_EXCEPTION = 2; // app code execution exception=
+        TRANSACTION_RESULT_VALIDATION_FAILED = 3; // TBD
```
```suggestion
        TRANSACTION_RESULT_VALIDATION_FAILED = 3; // TODO: do we need this? can an invalid tx ever make it into a superblock?
```

avive commented:
Renamed pr to api 2.0 as it is not going to be backward compatible with 1.x api due to some breaking changes.

lrettig commented:
> Renamed pr to api 2.0 as it is not going to be backward compatible with 1.x api due to some breaking changes.

I think we need to update this so the output proto files go into `proto/spacemesh/v2` (rather than `v1`), that should cause the CI tests to pass. We also need to add a separate set of checks for `v2`, and we should also update the README accordingly.
